### ISSUE TITLE
Support for server plaintext upgrade

### ISSUE THREAD

**Issue by dconnelly** [grpc/grpc-java#2228]
Looks like this is one of the supported protocol negotiators for clients but not for servers.

I've been trying to set up plaintext GRPC endpoints behind Amazon's new ALB (Application Load Balancer) and it looks like they require an upgradable connection on the backend (sadly, there is no way to configured ALB to assume HTTP/2).

I can try implementing a ProtocolNegotiator myself using Http2ServerUpgradeCodec but wanted to make sure this was feasible first.

**Comment by dconnelly** [comment]
As I suspected this is not so easy. Following a successful upgrade negotiation the full http request needs to be passed to the grpc transport for handling, and I can't see a way to do this using just a ProtocolNegotiator. I guess the other option would be use TLS on the backend but I was hoping to just let ALB handle the termination :(

**Comment by ejona86** [comment]
@dconnelly, yeah, the plain text upgrade is a bit of a pain. The easiest way to do it is to use a trash request initially, which is what we [do on client-side](https://github.com/grpc/grpc-java/blob/v1.0.0/netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java#L568). Since the initial request _must_ be completed before further usage of the connection, it is actually a bit wonky for a normal gRPC RPC since it could be a long-lived or streaming RPC. We could pretty easily fail the initial request on server-side, although ALB will probably send legit traffic initially which will cause spurious errors.
